### ISSUE TITLE
SNMP Library Support for New-SensorParameters discovery

### ISSUE THREAD

**Issue by JustinGrote** [lordmilko/PrtgAPI#69]
With the current PRTG version, Fiddler shows that a special syntax is required to get the dynamic parameters of SNMP Libraries.

https://myserver/controls/addsensor2.htm?id=8322&sensortype=snmplibrary_nolist&preselection_snmplibrary_nolist=ads.Talari.oidlib

(Note the preselection_snmplibrary_nolist)
Currently I don't see a way to specify the preselection parameter in the current PRTGAPI

Here's a simple proof of concept using httpClient in Powershell (couldn't get Invoke-Restmethod to work due to the additional unremovable headres it adds) for getting the raw HTML containing the table values.

```powershell
$deviceId = 8322
$baseaddress = 'https://myprtgserver'
$sensorType = 'snmplibrary_nolist'
$snmpLibraryName = 'ads.Talari.oidlib'
$username = 'myusername'
$passhash = '1133mypasshash445566'



#Couldn't get this to work with Invoke-Restmethod on PS5.1, dropping down to HTTPClient
$httpclient = [net.http.httpclient]::new()
$httpclient.baseaddress = $baseaddress
$requestUri = "controls/addsensor2.htm?id=$deviceID&sensortype=$sensorType&preselection_snmplibrary_nolist=$snmpLibraryName&username=$username&passhash=$passhash"
$addSensorResult = $httpclient.GetStringAsync($requestUri).Result

if ($addsensorResult -match 'name="tempid" value="(\d+)"') {
    $tmpid = $matches[1]
} else {
    throw "There was a problem with addsensor2 step"
}

do {
    sleep 0.5
    $targetUrl = ($httpclient.GetStringAsync("api/getaddsensorprogress.htm?id=$deviceID&tmpid=$tmpid").Result | ConvertFrom-Json).targeturl
} while (-not $targetUrl)

#Here's the HTML with the table to parse for the possible values to select
$httpclient.GetStringAsync($targetUrl).result
```

Example Table Entry
```
        <tr class=" notconnected  even">
            <td class="checkboxholder"><input type="checkbox" class="hidden validateme checkbox data-table"
                    data-rule-requiredtablecheckbox="true" class="checkbox validateme  notconnectedAll"
                    name="interfacenumber__check" id="interfacenumber_-20"
                    value="1.3.6.1.4.1.34086.2.2.16.2.1.4.2|TALARI-MIB|Wan Link  13: CAN-COMCAST1|State|#|0|0|State|2|1|0|1|The current operational state of the conduit.|0|0|0|0|oid.talari-mib.tnstatswanlink.tnstatswanlinkstate|1.3.6.1.4.1.34086.2.2.16.2.1.4|tnStatsWANLinkState|1.3.6.1.4.1.34086.2.2.16.2.1||ASN_INTEGER|0|ASN_INTEGER||TALARI-MIB|tn statswanlink: #[1.3.6.1.4.1.34086.2.2.16.2.1.1]|tn statswanlink state|||||||||||||||||||||||||||||||||||"><label
                    for="interfacenumber_-20"><i class="icon-dark prtg-checkbox"></i></label></td>
            <td title="">TALARI-MIB</td>
            <td title="">Wan Link 13: CAN-COMCAST1</td>
            <td title="">State</td>
        </tr>
```

### Implementation Proposal
Lots of different ways this could be done, @lordmilko I invite feedback if you've done any existing work on this and have any plans, otherwise I will write a Powershell function that will parse the table with HTMLAgilityPack, and produce a SensorParameters that can be passed to add-sensor and make it available as a Gist or something. The list of OIDLIB files is in /api/sensortypes.json so I would make it query that for an autocompleter.

**Comment by lordmilko** [comment]
Hi @JustinGrote,

As noted on the wiki, PrtgAPI does not currently support retrieving sensor targets or dynamic sensor parameters for sensors that you require you to enter some information before PRTG will load the sensor creation page. SNMP Library sensors fall into this category, and as such can only be created by creating a set of [custom parameters](https://github.com/lordmilko/PrtgAPI/wiki/Custom-Parameters).

I have had a thought about this previously and believe this external interface for this feature can be implemented by allowing the user to specify a list or dictionary or something to `GetSensorTargets` / `GetDynamicSensorParameters`. In the PowerShell interface this would presumably be achieved via parameter accepting a hashtable or something. These *query parameters* would then be passed to `BeginAddSensorQueryParameters`, where they would be added as a set `CustomParameter` objects

While implementing the external user interface is fairly straight forward to implement, what is not clear is how to implement a system to *tell you what query parameters should be specified when you fail to enter them properly*. Is that even possible? What is the page that is opened to select a SNMP Library when you go to to add the sensor? How does PRTG know that page needs to be opened? What requests are made? These questions will require a bit of investigating to determine the best way to implement the feature.

My current priority is getting .NET Core/.NET Standard support shipped, I will add this issue to the roadmap to hopefully tackle after .NET Core support is implemented

Regards,
lordmilko

**Comment by JustinGrote** [comment]
I agree that is priority. I'll write a wrapper for now. Thanks!

**Comment by JustinGrote** [comment]
Here's the proof of concept
https://gist.github.com/JustinGrote/d39ce1c1f398aa53354fdbc0b46b3049

### Contributes

- **Get-PRTGSNMPLibraryTargets** cmdlet that autocompletes with the list of available oidlibs on the server, and then you can use that along with a sensorparameters

- **Add-PRTGSNMPLibrarySensor** cmdlet that takes the library targets and the standard snmplibrary New-Sensorparameters output and adds it. I used [httpclient] for this one because one of the parameters repeatsand invoke-restmethod doesn't support that.

### Quirks Learned

- tmpid is optional for the creation, so is the session cookie, it can be done effectively stateless
- both the webgui and windows client use a multipart form to submit the adddsensor5.htm, but a GET query works just as well
- you have to use name and not name_ for this creation, so I couldn't use Add-Sensor because the sensorparameters requires name_ and that causes it to fail. **Maybe add a switch option for Add-Sensor for compatability?**

Is there any way to get at the httpclient that the PRTGClient uses? Maybe expose it as a public property?

**Comment by lordmilko** [comment]
Thanks @JustinGrote,

I'm not sure what the rules are like in the current version of PRTG and/or for SNMP Library sensors, however generally speaking the `tmpid` and `cookie` are definitely *not* optional. In my experience, failure to maintain the correct state when performing sensor creation queries will result in PRTG responding that a session is no longer valid. This is definitely true of sensors that take a long time to query; if you don't identify your session, how will PRTG know what progress percentage your request is up to? The only page that is stateless is the page that actually creates the sensor, `addsensor5.htm`

I can see you're right - SNMP Library sensors do appear to utilize a parameter called `name`, not `name_`. I'll need to figure out an appropriate way to implement this in a way that works for both dynamic sensor parameters and raw parameters.

PrtgAPI currently supports the `api/sensortypes.json` endpoint through the `Get-SensorType` cmdlet. Currently however it only deserializes the bare minimum of properties. From a quick check I can see that calling this page is the technique PRTG uses itself, so I will need to expand the properties of the `SensorTypeDescriptor` type.

The fact PrtgAPI uses a `HttpClient` is an implementation detail. My `HttpClient` doesn't really do anything that yours doesn't do. The only part of PrtgAPI that even knows a `HttpClient` is the [class](https://github.com/lordmilko/PrtgAPI/blob/master/PrtgAPI/Request/PrtgWebClient.cs) the implements the [IWebClient](https://github.com/lordmilko/PrtgAPI/blob/master/PrtgAPI/Request/IWebClient.cs) interface. If PrtgAPI's existing extension mechanisms (supporting "raw" values, etc) do not support a given feature, users can either invoke the request directly via `Invoke-WebRequest`, etc as a one off, or modify the PrtgAPI source code specifically for their use case.

I can see that the complexity of implementing this feature is quickly blowing up; I'm glad I made the decision to put a pin in this for now! :P

Regards,
lordmilko

**Comment by JustinGrote** [comment]
Yes it's not a simple fix. My main concern of getting at your HttpClient is
that it maintains the state and cookies, and you can use that instead of
constantly passing username and passhash, so I can run commands "in module
scope" of sorts without having to instantiate a new session.

In regards to tmpid you'll see I do preserve it all the way through
addsensor thorugh addsensor4, my point was just that it doesn't appear you
need it for the addsensor5 step.

The sensortypes.json offers a boatload of information about sensors so it's
a good starting point for building parameters, hence why I made the
autocomplete as a proof of concept.

100% agree core comes first, I just had a pressing need for this (need to
onboard about 500 snmp library sensors for Talari systems that all have
different interface indexes and some need to be filtered), so I just wanted
to check that I wasn't reinventing the wheel.

**Comment by lordmilko** [comment]
Hi @JustinGrote,

I am currently doing some research into the implementation of this feature.

Are you able to advise whether you are aware of any other sensor types that give initial prompts in the PRTG UI prior to taking you to the normal sensor creation page?

So far, I am aware of the following types

| SensorType         | Stage                            | Content       | Mandatory |
| ------------------ | ------------------------- | ------------- | ------------ |
| snmplibrary         | After addsensor2.htm  | Preselection | Yes            |
| oracletablespace | After addsensor3.htm  | Partial query | No            |

Depending on the types of information you can get away with not specifying during the requests leading up to the normal sensor creation page, this will determine how simple or complicated the API for this will have to be

Regards,
lordmilko

**Comment by JustinGrote** [comment]
I'm not aware of any outside of SNMP Library off the top of my head, short
of taking the time to click every sensor and find out :/

**Comment by lordmilko** [comment]
Hi @JustinGrote,

Please be advised that PrtgAPI 0.9.7 has now been released, which includes fixes for these issues

To update to the latest version, run

```powershell
Update-Module PrtgAPI
```

and reopen PowerShell. Please see the [release notes](https://github.com/lordmilko/PrtgAPI/releases/tag/v0.9.7) for a list of breaking changes that may affect you.

PrtgAPI 0.9.7 introduces the concept of *sensor query targets*, which unlike *sensor targets*, merely specify the target that a *query* should apply to (such as the oidlib to use for an SNMP Library, the details of a database server or the authentication details to use for a niche sensor type). For more information on sensor query targets, please see the [wiki](https://github.com/lordmilko/PrtgAPI/wiki/Objects#sensor-query-targets).

In your case, you should now be able to create a sensor for the oidlib `ads.Talari.oidlib` via the following code

```powershell
$device = Get-Device -Id 8322
$params = $device | New-SensorParameters -RawType snmplibrary -qt "ads.Talari.oidlib"
$device | Add-Sensor $params
```

Attempting to query dynamic sensor parameters or sensor targets of a sensor type that requires some type of sensor query target will now generate an `InvalidOperationException`, specifying the possible values that can be specified.

`SensorTypeDescriptor` objects returned from `Get-SensorType` now include a `QueryTargets` property that can be used to retrieve `SensorQueryTarget` objects, however generally it is a lot simpler to specify a wildcard value to the new `-QueryTarget` parameter.

For sensor types that require multiple values be specified, you can do so using the new `-QueryParameter` property

```powershell
$params = $device | New-SensorParameters -RawType ipmisensor -qp @{
    username = "admin"
    password = "password"
}
```

For sensor types that require custom parameter names be used for common parameters defined under `NewSensorParameters` / `NewObjectParameters` PrtgAPI now allows you to *override* these parameter names with something different. For more information, see the [wiki](https://github.com/lordmilko/PrtgAPI/wiki/Custom-Parameters#overriding-parameters). Note however that when retrieving `DynamicSensorParameters`, PrtgAPI will now *automatically apply any required name overrides for you* when it detects a typed dynamic parameter name is different from the name it would use by default.

Thanks for raising this issue!

Regards,
lordmilko

**Comment by JustinGrote** [comment]
@lordmilko I apologize I got pulled off to another priority and didn't get to test this. 

Initially the parameter discovery works great and I get all the relevant OIDs for the SNMPLibrary
```powershell
$sensorparameters = new-sensorparameters -RawType snmplibrary -QueryTarget 'ads.Talari.oidlib' -device (get-device -id 8322) -verbose
$sensorparameters.targets.interfacenumber__check
```

```
Value                             Properties                                                                                                 Name
-----                             ----------                                                                                                 ----
1.3.6.1.4.1.34086.2.2.17.2.1.12.2 {1.3.6.1.4.1.34086.2.2.17.2.1.12.2, TALARI-MIB, Conduit 458795: NORTHCHASE_DR-CANDLER-GA, Jitter…}         TALARI-MIB    
1.3.6.1.4.1.34086.2.2.17.2.1.9.2  {1.3.6.1.4.1.34086.2.2.17.2.1.9.2, TALARI-MIB, Conduit 458795: NORTHCHASE_DR-CANDLER-GA, Traffic Dropped…} TALARI-MIB
1.3.6.1.4.1.34086.2.2.17.2.1.7.2  {1.3.6.1.4.1.34086.2.2.17.2.1.7.2, TALARI-MIB, Conduit 458795: NORTHCHASE_DR-CANDLER-GA, Traffic In…}      TALARI-MIB
1.3.6.1.4.1.34086.2.2.17.2.1.5.2  {1.3.6.1.4.1.34086.2.2.17.2.1.5.2, TALARI-MIB, Conduit 458795: NORTHCHASE_DR-CANDLER-GA, Traffic Out…}     TALARI-MIB
1.3.6.1.4.1.34086.2.2.17.2.1.4.2  {1.3.6.1.4.1.34086.2.2.17.2.1.4.2, TALARI-MIB, Conduit 458795: NORTHCHASE_DR-CANDLER-GA, State…}           TALARI-MIB
1.3.6.1.4.1.34086.2.2.17.2.1.2.2  {1.3.6.1.4.1.34086.2.2.17.2.1.2.2, TALARI-MIB, Conduit 458795: NORTHCHASE_DR-CANDLER-GA, ID…}              TALARI-MIB
1.3.6.1.4.1.34086.2.2.17.2.1.12.1 {1.3.6.1.4.1.34086.2.2.17.2.1.12.1, TALARI-MIB, Conduit 7: INTERNAP-COLO-CANDLER-GA, Jitter…}              TALARI-MIB    
1.3.6.1.4.1.34086.2.2.17.2.1.9.1  {1.3.6.1.4.1.34086.2.2.17.2.1.9.1, TALARI-MIB, Conduit 7: INTERNAP-COLO-CANDLER-GA, Traffic Dropped…}      TALARI-MIB    
1.3.6.1.4.1.34086.2.2.17.2.1.7.1  {1.3.6.1.4.1.34086.2.2.17.2.1.7.1, TALARI-MIB, Conduit 7: INTERNAP-COLO-CANDLER-GA, Traffic In…}           TALARI-MIB    
1.3.6.1.4.1.34086.2.2.17.2.1.5.1  {1.3.6.1.4.1.34086.2.2.17.2.1.5.1, TALARI-MIB, Conduit 7: INTERNAP-COLO-CANDLER-GA, Traffic Out…}          TALARI-MIB
1.3.6.1.4.1.34086.2.2.17.2.1.4.1  {1.3.6.1.4.1.34086.2.2.17.2.1.4.1, TALARI-MIB, Conduit 7: INTERNAP-COLO-CANDLER-GA, State…}                TALARI-MIB
1.3.6.1.4.1.34086.2.2.17.2.1.2.1  {1.3.6.1.4.1.34086.2.2.17.2.1.2.1, TALARI-MIB, Conduit 7: INTERNAP-COLO-CANDLER-GA, ID…}                   TALARI-MIB    
1.3.6.1.4.1.34086.2.2.16.2.1.9.1  {1.3.6.1.4.1.34086.2.2.16.2.1.9.1, TALARI-MIB, Wan Link  12: CAN-CENTURYLINK1, Traffic Dropped…}           TALARI-MIB    
1.3.6.1.4.1.34086.2.2.16.2.1.7.1  {1.3.6.1.4.1.34086.2.2.16.2.1.7.1, TALARI-MIB, Wan Link  12: CAN-CENTURYLINK1, Traffic In…}                TALARI-MIB    
1.3.6.1.4.1.34086.2.2.16.2.1.5.1  {1.3.6.1.4.1.34086.2.2.16.2.1.5.1, TALARI-MIB, Wan Link  12: CAN-CENTURYLINK1, Traffic Out…}               TALARI-MIB
1.3.6.1.4.1.34086.2.2.16.2.1.4.1  {1.3.6.1.4.1.34086.2.2.16.2.1.4.1, TALARI-MIB, Wan Link  12: CAN-CENTURYLINK1, State…}                     TALARI-MIB    
1.3.6.1.4.1.34086.2.2.16.2.1.2.1  {1.3.6.1.4.1.34086.2.2.16.2.1.2.1, TALARI-MIB, Wan Link  12: CAN-CENTURYLINK1, ID…}                        TALARI-MIB    
1.3.6.1.4.1.34086.2.2.16.2.1.9.2  {1.3.6.1.4.1.34086.2.2.16.2.1.9.2, TALARI-MIB, Wan Link  13: CAN-COMCAST1, Traffic Dropped…}               TALARI-MIB    
1.3.6.1.4.1.34086.2.2.16.2.1.7.2  {1.3.6.1.4.1.34086.2.2.16.2.1.7.2, TALARI-MIB, Wan Link  13: CAN-COMCAST1, Traffic In…}                    TALARI-MIB
1.3.6.1.4.1.34086.2.2.16.2.1.5.2  {1.3.6.1.4.1.34086.2.2.16.2.1.5.2, TALARI-MIB, Wan Link  13: CAN-COMCAST1, Traffic Out…}                   TALARI-MIB    
1.3.6.1.4.1.34086.2.2.16.2.1.4.2  {1.3.6.1.4.1.34086.2.2.16.2.1.4.2, TALARI-MIB, Wan Link  13: CAN-COMCAST1, State…}                         TALARI-MIB    
1.3.6.1.4.1.34086.2.2.16.2.1.2.2  {1.3.6.1.4.1.34086.2.2.16.2.1.2.2, TALARI-MIB, Wan Link  13: CAN-COMCAST1, ID…}                            TALARI-MIB
```

However, in attempting to create the sensor itself, it only creates the first channel of the first sensor. What is expected is to create 4 interface sensors with individual channels each (22 channels total) as my powershell above and the GUI do. I tried creating it in this fashion:
`get-device -id 8322 | add-sensor -parameters $sensorparameters`

Let me know if the syntax is correct or if I need to go about it otherwise. I also would assume if I filtered the targets property and only wanted to create a couple of the OIDs, that it would work as expected if I edited that property.

**Comment by lordmilko** [comment]
Hi @JustinGrote,

By default only the first sensor target is selected. Therefore you need to add `-Target *` to your call to `New-SensorParameters` (or filter the list above manually yourself before assigning it)

Regards,
lordmilko

**Comment by JustinGrote** [comment]
@lordmilko that was it. Out of curiosity, why is * not the default for -Target?

**Comment by JustinGrote** [comment]
@lordmilko 
So the -Target appears to be a string filter on the name property of the GenericSensorTarget objects that are returned. The problem is that, in the case of SNMP library, they are all the same.
```
new-sensorparameters -rawtype snmplibrary -device $device -querytarget 'ads.talari.oidlib' -target '*' | % interfacenumber__check | select name,value

Name       Value
----       -----
TALARI-MIB 1.3.6.1.4.1.34086.2.2.17.2.1.12.2
TALARI-MIB 1.3.6.1.4.1.34086.2.2.17.2.1.9.2
TALARI-MIB 1.3.6.1.4.1.34086.2.2.17.2.1.7.2
TALARI-MIB 1.3.6.1.4.1.34086.2.2.17.2.1.5.2
TALARI-MIB 1.3.6.1.4.1.34086.2.2.17.2.1.4.2
TALARI-MIB 1.3.6.1.4.1.34086.2.2.17.2.1.2.2
TALARI-MIB 1.3.6.1.4.1.34086.2.2.17.2.1.12.1
TALARI-MIB 1.3.6.1.4.1.34086.2.2.17.2.1.9.1
TALARI-MIB 1.3.6.1.4.1.34086.2.2.17.2.1.7.1
TALARI-MIB 1.3.6.1.4.1.34086.2.2.17.2.1.5.1
TALARI-MIB 1.3.6.1.4.1.34086.2.2.17.2.1.4.1
TALARI-MIB 1.3.6.1.4.1.34086.2.2.17.2.1.2.1
TALARI-MIB 1.3.6.1.4.1.34086.2.2.16.2.1.9.1
TALARI-MIB 1.3.6.1.4.1.34086.2.2.16.2.1.7.1
TALARI-MIB 1.3.6.1.4.1.34086.2.2.16.2.1.5.1
TALARI-MIB 1.3.6.1.4.1.34086.2.2.16.2.1.4.1
TALARI-MIB 1.3.6.1.4.1.34086.2.2.16.2.1.2.1
TALARI-MIB 1.3.6.1.4.1.34086.2.2.16.2.1.9.2
TALARI-MIB 1.3.6.1.4.1.34086.2.2.16.2.1.7.2
TALARI-MIB 1.3.6.1.4.1.34086.2.2.16.2.1.5.2
TALARI-MIB 1.3.6.1.4.1.34086.2.2.16.2.1.4.2
TALARI-MIB 1.3.6.1.4.1.34086.2.2.16.2.1.2.2
```

So the target parameter is useless in this scenario and should allow for a richer evaluation.

I was able to get it to work by unlocking the returned object and filtering out the unwanted targets, but obviously unlocking is an "off the reservation activity"
```powershell
$sp = new-sensorparameters -rawtype snmplibrary -device $device -querytarget 'ads.talari.oidlib' -target '*'
$sp.unlock()
$sp.interfacenumber__check = $sp.interfacenumber__check.where{$PSItem.properties[2] -notmatch 'Centurylink'}

**Comment by lordmilko** [comment]
As per the [wiki](https://github.com/lordmilko/PrtgAPI/wiki/Dynamic-Parameters#targets-1), you can either filter by `-Target` or filter the targets directly yourself once the parameters have been returned

```powershell
# Untested but more or less correct
$params = new-sensorparameters -rawtype snmplibrary -device $device -querytarget 'ads.talari.oidlib'
$params.interfacenumber__check = $params.Targets.interfacenumber__check | where { $_.properties[2] -notmatch 'Centurylink' }
```

**Comment by lordmilko** [comment]
In regards to your query (now deleted?) the `DynamicSensorParameters` type constructs itself by
1. Instantiating its parameters using their default values, and then
2. Constructing a dictionary of values that appear to be suitable as sensor targets

As a result of this, any property that *does* represent a sensor target will only have the first sensor target in the list. While I can think of a number of scenarios where selecting all targets by default would be useful (WMI Volumes, VMware Datastores, Exchange Databases, etc) there are other scenarios where it wouldn't be (WMI Services - you sure don't want a sensor for all 200 services on your system!)

Bottom line, for now I have decided not to go the additional step of selecting all of the targets by default, and have left it up to users to specify what they want. If you really do want to select all targets, it's not that hard to add `-Target * to your call to `New-SensorParamters`.

In your case however, you *don't* want to specify `-Target *` you should simply filter the targets via the `Targets` property of the parameters object *after* the object has been created

**Comment by JustinGrote** [comment]
@lordmilko I deleted it as I sorted it out but I appreciate the
explanation.
